rholang/src/main/scala/coop/rchain/rholang/interpreter/Registry.scala

```diff
@@ -0,0 +1,408 @@
+package coop.rchain.rholang.interpreter
+
+import com.google.protobuf.ByteString
+import com.google.protobuf.ByteString.ByteIterator
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.crypto.hash.Blake2b512Random
+import coop.rchain.models.Channel.ChannelInstance
+import coop.rchain.models.Channel.ChannelInstance.{ChanVar, Quote}
+import coop.rchain.models.Expr.ExprInstance
+import coop.rchain.models.Expr.ExprInstance._
+import coop.rchain.models.TaggedContinuation.TaggedCont.ScalaBodyRef
+import coop.rchain.models.Var.VarInstance.FreeVar
+import coop.rchain.models._
+import coop.rchain.models.rholang.implicits._
+import coop.rchain.rholang.interpreter.storage.implicits._
+import coop.rchain.rspace.{ISpace, IStore}
+import monix.eval.Task
+import scala.collection.immutable.Seq
+import scala.collection.{Seq => RootSeq}
+
+class Registry(private val space: ISpace[Channel,
+                                         BindPattern,
+                                         ListChannelWithRandom,
+                                         ListChannelWithRandom,
+                                         TaggedContinuation],
+               private val dispatcher: Dispatch[Task, ListChannelWithRandom, TaggedContinuation]) {
+  import Registry._
+  def commonPrefix(b1: ByteString, b2: ByteString): ByteString = {
+    val prefixOut = ByteString.newOutput()
+    def loop(it1: ByteIterator, it2: ByteIterator): ByteString =
+      if (!it1.hasNext || !it2.hasNext) {
+        prefixOut.toByteString
+      } else {
+        val b: Byte = it1.nextByte
+        if (it2.nextByte == b) {
+          prefixOut.write(b.toInt)
+          loop(it1, it2)
+        } else {
+          prefixOut.toByteString()
+        }
+      }
+    loop(b1.iterator, b2.iterator)
+  }
+
+  def safeUncons(b: ByteString): (ByteString, ByteString) = {
+    val head = if (b.isEmpty()) b else b.substring(0, 1)
+    val tail = if (b.isEmpty()) b else b.substring(1)
+    (head, tail)
+  }
+
+  val lookupRef: Long = 10L
```

Comment:
We need to find a way to prevent from `SystemProcesses` overwriting these. Maybe pass initial value that is one bigger that the biggest ref value in `SystemProcesses` ?